### ISSUE TITLE
Server: backend sampling (-bs) gives +48% throughput at 32 slots (706 -> 1046 tok/s) — should it be the default for multi-slot serving?

### ISSUE THREAD

**Issue by swiftjiang** [ggml-org/llama.cpp#27050]
### Summary

<img width="1690" height="676" alt="Image" src="https://github.com/user-attachments/assets/89b4ac18-95eb-4fc1-be78-d3e3b47f78b1" />

While benchmarking `llama-server` at high concurrency on a single RTX 5090, I found that
generation throughput saturates around **700 tok/s** with 32 slots, and that Nsight Systems
attributes this to the server copying the **entire logits matrix device-to-host on every
decode step** (16.2 MB per step at batch 32).

CPU sampling confirms the other half of the loop: inside those GPU-idle windows, **45% of
the application's on-CPU samples are in `common_sampler_sample()`** and another 13% in
`llama_token_data_array_partial_sort_inplace()`.

The `-bs` / `--backend-sampling` flag (merged very recently in #25532) removes that transfer
completely and raises throughput to **~1046 tok/s (+48%)**, with TPOT dropping from 41.2 ms
to 25.4 ms, and the sampler's share of in-gap CPU time falling from 58.6% to 1.0%.
Single-slot performance is unchanged, and greedy output is byte-identical between the two
paths.

Since `-bs` is currently experimental and off by default, I wanted to post the multi-slot
numbers, because the PR that introduced it reported ~8% (single-user, speculative decoding)
and I could not find any published measurement of its effect on a many-slot server.

Two questions at the end: whether `-bs` is expected to become the default for multi-slot
serving, and what accounts for the GPU idle time that still remains after enabling it.

### Environment

- llama.cpp: `a94d563ed801d1da1b8c2432946de07d0231bb3d` (build 10423), CUDA backend
  - also verified against `7dad2f1a1` (build 9660, 2026-06-15) — numbers are within 1%,
    so the default-path behaviour has not changed over the last ~760 commits
- Build: `cmake -B build -DGGML_CUDA=ON -DCMAKE_CUDA_ARCHITECTURES=120 -DLLAMA_CURL=OFF -DCMAKE_BUILD_TYPE=Release`
- GPU: NVIDIA GeForce RTX 5090 32GB (Blackwell, sm_120), driver 595.84
- CPU: Intel Xeon E5-2673 v4 @ 2.30GHz, 40 cores (2016-era, modest single-core performance)
- RAM: 31 GB, OS: Ubuntu 24.04.2, CUDA Toolkit 12.9, Nsight Systems 2025.1.3
- Model: Qwen2.5-7B-Instruct, F16 GGUF (14.19 GiB), converted from the official HF weights
- Reference engine: vLLM, same GPU, same model, BF16, same client and workload

### Server command

```
llama-server -m Qwen2.5-7B-Instruct-F16.gguf \
  -ngl 99 -c 32768 -np 32 -cb -fa on --ignore-eos --cache-ram 0 \
  --host 127.0.0.1 --port 8080 --threads-http 32
```

The `-bs` run is the same command with `-bs` appended, nothing else changed.

### Workload

Every request is exactly 512 prompt tokens + exactly 128 generated tokens, greedy
(`temperature=0`), `ignore_eos`. Prompts are sent as **token ID arrays** so the input is
byte-identical across engines. Each request carries a unique prefix and the prompt cache is
disabled (`--cache-ram 0`), so every request performs a real prefill. Closed loop: N workers
each keep one request in flight. Measurements are client-side, 2 repetitions, mean reported.

### Results

| Configuration | 1 slot | | 32 slots | |
|---|---:|---:|---:|---:|
| | tok/s | TPOT | tok/s | TPOT |
| `7dad2f1a1` (build 9660), default | 99.8 | 9.67 ms | 700.9 | 41.50 ms |
| `a94d563ed` (build 10423), default | 100.1 | 9.65 ms | 705.9 | 41.21 ms |
| `a94d563ed` (build 10423), **`-bs`** | 101.4 | 9.43 ms | **1045.8** | **25.39 ms** |
| vLLM BF16 (reference) | 92.9 | 10.41 ms | 1404.7 | 17.68 ms |

Per-repetition values at 32 slots: default 708.1 / 703.7 tok/s, `-bs` 1039.8 / 1051.9 tok/s.

Note that llama.cpp is *faster* than vLLM at 1 slot, so this is specifically about
concurrency scaling, not general performance. The vLLM column is there to show the hardware
still has headroom under this exact workload, not as a quality judgement — the two projects
have different design goals.

### Nsight evidence: what `-bs` actually removes

Profiled with `nsys profile --trace=cuda,osrt --cuda-graph-trace=node`, 10 s window under
steady-state load. Server-side TPOT during the traced window was 42.2 ms (default) and
27.6 ms (`-bs`), i.e. close to the untraced values, so the traces are representative.

Device-to-host memcpy volume, 32 slots:

| Configuration | D2H copies | Total | **Per copy** |
|---|---:|---:|---:|
| default | 181 | 2.924 GB | **16157.4 KB** |
| **`-bs`** | 16252 | ~0.000 GB | **0.0 KB** |

At 1 slot the default path transfers **608.3 KB** per copy, which is exactly
`152064 vocab x 4 B = 608,256 B` — one complete logits row — and it happens once per decode
step (938 copies in a 10 s window at ~94 decode steps/s). At 32 slots the same per-step copy
becomes ~16 MB. With `-bs` the payload becomes sampled token IDs and the transferred bytes
round to zero.

So the mechanism is directly measured, not inferred: the default server path moves
`batch x n_vocab x 4 B` to the host every decode step, and that is what `-bs` eliminates.

### GPU utilization

Default path, same traces:

| Configuration | GPU busy | Gaps >= 1 ms |
|---|---:|---|
| default, 1 slot | 88.5% | 27 gaps, 0.15 s / 10 s |
| default, 32 slots | 41.2% | 296 gaps, 5.70 s / 10 s (mean 19.3 ms) |

GPU utilization *decreases* as concurrency increases, which is the observation that started
this investigation.

I am deliberately **not** quoting a GPU-busy figure for the `-bs` run. In that
configuration `cudaGraphLaunch` goes from 144 to 502 calls in the same window, so much more
of the work runs inside CUDA graphs, and I do not trust the node-level trace to cover both
configurations equally (the traced kernel count comes out lower than the decode-step count
implies). Throughput, TPOT and memcpy volume are all direct measurements and are not
affected by this.

### Host-side hotspot (CPU sampling)

Separate runs with `--sample=process-tree --cpuctxsw=process-tree` (`perf_event_paranoid=1`).
Nsight 2025.1.3 has no `cpu_sample_sum` report, so these are taken straight out of the
`COMPOSITE_EVENTS` / `SAMPLING_CALLCHAINS` tables in the exported SQLite, counting leaf
(`stackDepth = 0`) symbols of samples whose timestamp falls inside a >= 1 ms GPU gap.
Kernel-mode samples and Nsight's own injection libraries are excluded so that the
percentages describe application code only.

32 slots, samples landing inside GPU-idle gaps:

| Self function | default | `-bs` |
|---|---:|---:|
| `common_sampler_sample(common_sampler*, llama_context*, int, bool)` | **45.2%** | – |
| `llama_token_data_array_partial_sort_inplace(...)` | **12.7%** | – |
| **sampler total** | **58.6%** | **1.0%** |
| `ggml_backend_sched_split_graph` | 0.1% | 2.4% |
| `llama_kv_cache::prepare(...)` | 0.5% | 1.8% |
| mutex + malloc/free primitives (`libc`) | 9.9% | 24.6% |

(10079 in-gap application samples for the default run, 4934 for `-bs`.) At 1 slot the same
measurement puts the sampler at 5.1% of a much smaller idle budget (0.185 s in 10 s).

So the host-side cost that the D2H transfer feeds into is the sampler itself, and `-bs`
removes essentially all of it. What is left in the `-bs` run is spread across graph
scheduling, KV-cache preparation and allocator/mutex traffic, with no single dominant entry.

### Correctness check

With `temperature=0`, `seed=42` and a fixed prompt, the default and `-bs` paths produce
**byte-identical** output, so the speedup is not coming from a degraded path.

### What I ruled out on the default path

These control experiments were run on `7dad2f1a1`, not re-run on `a94d563ed`; since the
default path reproduces within 1% across the two builds I believe they still apply. Each
tested at 32 slots; none changed throughput by more than 10%:

| Hypothesis | Test | Result |
|---|---|---|
| Flash attention config | `-fa on` vs unspecified | 717 vs 713 tok/s |
| Internal perf timers | `--no-perf` | no change |
| Full-vocab top-k sort in sampler | request with `top_k=1` | no change (711) — see note |
| Streaming/SSE overhead | non-streaming requests | 787 tok/s (+9.7%) |
| Too few slots | `-np 64 -c 65536` | +7%, still saturated |
| Too few HTTP threads | `--threads-http 32` and `64` | no change |
| Machine/thermal degradation | clocks at 2887 MHz, no throttle flags | ruled out |
| Client being the bottleneck | same client drives vLLM to 2x throughput | ruled out |

Note on the `top_k=1` row: setting `top_k=1` does not help, yet the profile still shows 12.7%
of in-gap CPU time in `llama_token_data_array_partial_sort_inplace`. Those are consistent —
the per-step cost is dominated by `common_sampler_sample` walking the full `n_vocab` array
regardless of `top_k`, so tightening `top_k` cannot avoid it. Only moving the work off the
host (`-bs`) does.

### The remaining gap

Even with `-bs`, 1046 tok/s is still below vLLM's 1405 tok/s on this hardware, and gaps
>= 1 ms still account for roughly 6 s of a 10 s window. CUDA API calls only explain ~11% of
that gap time; the rest has no CUDA activity at all. Two things I noticed that may or may
not be related:

- `cudaGraphLaunch` spends 385–449 ms *inside* GPU-idle windows (~7% of gap time) in both
  configurations, ~2.7 ms per call.
- `pthread_mutex_lock` accounts for ~307% of gap time at 32 slots (i.e. ~3 threads blocked
  on average) versus 0.1% at 1 slot, so there is lock contention that scales with concurrency.
- In the `-bs` CPU-sampling run, mutex and allocator primitives together are 24.6% of in-gap
  application samples (versus 9.9% on the default path), so once the sampler is out of the
  way, lock and allocator traffic become the largest identifiable application-side cost.

Separately, `llama-server` has no chunked prefill (cf. #10718), so a 512-token prefill
occupies a full forward pass and stalls decoding for other slots. Sending 8-token prompts
instead of 512-token prompts at 32 slots raised default-path throughput from ~700 to
957 tok/s (measured on `7dad2f1a1`), so prefill interference accounts for part of the
remainder.

### Limitations

- The CPU-sampling runs carry visible Nsight overhead (its injection libraries account for
  ~16% of all on-CPU samples), which is why the hotspot percentages are computed over
  application code only, and why GPU-busy figures are quoted from the non-sampling runs.
- **Nsight caveat worth repeating**: `--cuda-graph-trace=node` is required. With the default
  `graph` granularity, kernels inside CUDA graphs are not traced at all and GPU utilization at
  1 slot comes out as 4.7% instead of 88.5%.
- The CPU here is a 2016-era Xeon with modest single-core performance. A host-side bottleneck
  will be more visible here than on a modern CPU, so I would not extrapolate the *magnitude*
  of the `-bs` speedup to other machines.
- Single GPU, single model, single request shape (512 + 128). Not tested: variable-length
  requests, long contexts, long generations, non-greedy sampling.

### Reproduction

Benchmark client and profiling scripts (stdlib + aiohttp + transformers):

- https://gist.github.com/swiftjiang/187500e5b77f14b5080d1f96b18f944b

The Nsight reports (`.nsys-rep`), the per-run gap-attribution dumps and the CPU-hotspot
tables are all available on request — I did not attach them here to keep the issue
readable, but I can post any of them in a comment.

### Questions

1. Is `-bs` / `--backend-sampling` expected to become the default, at least when `-np > 1`?
   On this setup it is a 48% throughput win at 32 slots with identical greedy output, which
   seems like a large gain to leave behind an experimental opt-in flag. If it stays opt-in,
   would it be worth mentioning in `tools/server/README.md` that it matters for multi-slot
   serving?
2. With `-bs` enabled the sampler is gone from the profile, but the GPU is still idle for a
   large fraction of the window and no single function dominates what remains — it is spread
   across `ggml_backend_sched_split_graph`, `llama_kv_cache::prepare` and allocator/mutex
   traffic. Is that expected per-step scheduling overhead at 32 slots, or is this the prefill
   blocking described in #10718? Happy to collect whatever profile would be most useful.

Happy to run further experiments on this setup.
